typed-racket-lib/typed-racket/logic/ineq.rkt

```diff
@@ -0,0 +1,920 @@
+             hash-eq?)
+      lexp?)
+  (cons const terms))
+
+(define lexp-const car)
+(define lexp-terms cdr)
+
+(define-syntax-rule (lexp* t ...)
+  (list->lexp (list t ...)))
+
+(define/cond-contract (list->lexp terms)
+  (-> (listof (or/c exact-integer? (list/c exact-integer? any/c))) lexp?)
+  (let loop ([c 0]
+             [h (make-terms)]
+             [zxs terms])
+    (match zxs
+      [(list) (lexp c h)]
+      [(cons (list coeff var) zxs-rst)
+       (loop c (terms-set h var (+ coeff (terms-ref h var))) zxs-rst)]
+      [(cons constant zxs-rst)
+       (loop (+ constant c) h zxs-rst)])))
+
+; lexp-coeff
+(define/cond-contract (lexp-coeff l x)
+  (-> lexp? any/c exact-integer?)
+  (terms-ref (lexp-terms l) x))
+
+(define-syntax (check-true stx)
+  (syntax-case stx ()
+    [(_ exp) (syntax/loc stx (unless (eq? #t exp) (error 'check-true "test failed!")))]))
+(define-syntax (check-false stx)
+  (syntax-case stx ()
+    [(_ exp) (syntax/loc stx (when (not (eq? #f exp)) (error 'check-false "test failed!")))]))
+(define-syntax (check-not-false stx)
+  (syntax-case stx ()
+    [(_ exp) (syntax/loc stx (unless exp (error 'check-not-false "test failed!")))]))
+(define-syntax (check-equal? stx)
+  (syntax-case stx ()
+    [(_ val exp)
+     (syntax/loc stx (unless (equal? val exp)
+                       (error 'check-equal? "not equal!")))]))
+
+(module+ test
+
+  (check-equal? (lexp* 1 '(1 x) '(42 y) 1) (lexp* '(42 y) 2 '(1 x)))
+  (check-equal? (lexp* 0) (lexp* 0 '(0 x) '(0 y) '(0 z)))
+  (check-equal? (lexp-coeff (lexp* 1 '(1 x) '(42 y)) 'y) 42)
+  (check-equal? (lexp-const (lexp* 1 '(1 x) '(42 y))) 1)
+  (check-equal? (lexp-coeff (lexp* '(1 x) '(42 y)) 'q) 0))
+
+(define/cond-contract (lexp-vars exp)
+  (-> lexp? list?)
+  (terms-vars (lexp-terms exp)))
+
+(define/cond-contract (lexp-scalars exp)
+  (-> lexp? (listof exact-integer?))
+  (match exp
+    [(lexp: c ts)
+     (define coeffs (terms-coeffs ts))
+     (if (zero? c) coeffs (cons c coeffs))]
+    [_ (error 'lexp-scalars "given invalid lexp ~a" exp)]))
+
+(module+ test
+  (check-true (and (equal? (sort (lexp-vars (lexp* 17 '(42 x) '(2 z))) symbol<?)
+                           (sort '(x z) symbol<?))
+                   (= 2 (length (lexp-vars (lexp* 17 '(42 x) '(2 z)))))))
+  (check-true (and (equal? (sort (lexp-scalars (lexp* 17 '(42 x) '(2 z))) <)
+                           '(2 17 42))
+                   (= 3 (length (lexp-scalars (lexp* 17 '(42 x) '(2 z))))))))
+
+
+; lexp-scale
+;; if multiplying any scalar by a results
+;; in a non integer, error is thrown if
+;; contracts are active
+(define/cond-contract (lexp-scale exp a)
+  (-> lexp? rational? lexp?)
+  (cond
+    [(zero? a)
+     (lexp 0 (make-terms))]
+    [(= a 1) exp]
+    [else
+     (match exp
+       [(lexp: c h)
+        (lexp (* c a)
+              (terms-scale h a))])]))
+
+(module+ test
+  (check-equal? (lexp-set (lexp* 17 '(42 x)) 'x 0)
+                (lexp* 17)))
+; lexp-set
```

Comment:
personal preference (I like the different color flagging function defs) -- I can add descriptions as well.